src/viper/trans.ml

```diff
@@ -336,7 +336,7 @@ and dec_field' ctxt d =
   match d.M.dec.it with
   (* type declarations*)
   | M.(TypD (typ_id, typ_binds, {note = T.Variant cons;_})) ->
-    ctxt, None, None, fun _ ->
```

Comment:
I agree that shadowing is better, but here you have changed semantic.
Before we was using `ctxt` from `dec_field' ctxt d` which is environment of declarations while `ctxt` from `func ctxt` is environment of initialization.
There are different. For example the first former one saw prefix of declarations, while the latter all of them.
If you change semantic, please explain it in each case.

Reply:
In the types cases `ctxt` is passed to the `tr_typ`. `tr_typ` requires it because it updates the tuple arities for the subsequent prelude generation.

I guess that when the `ctxt` was introduced in `tr_typ` one forgot to add `ctxt` to the lambda argument.

Reply:
Ah, okay I see. Here is the bug?

Reply:
It may shoot later. Right now `tr_typ` refers to the `ref` field, so, the value inside is the same.

Reply:
I think you are not right. I would say that type definition should interpreted under context of previous definitions, so postponed context was omitted on purpose.
We can ask @int-index

Reply:
Okay I think it's not really important